File: 4/priority-q/pq.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>


typedef struct Item {
  int value;
  struct Item *next;
} Item;


Item *makeItem(int value) {
  Item *item = malloc(sizeof(Item));
  item->value = value;
  return item;
}

Item *enqueue(Item *head, int value) {
  if (head->value < value) {
    Item *item = makeItem(value);
    item->next = head;
    return item;
  }

  Item *current = head;

  while (current) {
    
    if (current->next && current->next->value < value) {
      Item *item = makeItem(value);
      item->next = current->next;
      current->next = item;
      return head;
    }

    if (!current->next) {
      Item *item = makeItem(value);
      item->next = current->next;
      current->next = item;
      return head;
    }
    
    current = current->next;
  }

  return head;
}

Item *dequeue(Item *head, int value) {
  return head -> next;
}


void printList(Item *head) {
  Item *item = head;
  while(item) {
    printf("%d\n", item->value);
    item = item->next;
  }
}

void main() {
  Item *item = makeItem(23);
  item = enqueue(item, 45);
  item = enqueue(item, 15);
  item = enqueue(item, 5);
  item = enqueue(item, 89);
  item = enqueue(item, 14);
  item = enqueue(item, 16);
  printList(item);
}

File: 20-short-path/pq.js
function PQ(cmp) {
  let head;

  function makeItem(value) {
    return {
      value,
      next: null
    }
  }

  function enqueue(value) {
    if (!head) {
      head = makeItem(value)
      return head;
    }

    if (cmp(head.value, value)) {
      const item = makeItem(value);
      item.next = head;
      head = item;
      return item;
    }

    let current = head;

    while (current) {
      if (current.next && cmp(current.next.value, value)) {
        const item = makeItem(value);
        item.next = current.next;
        current.next = item;
        return head;
      }

      if (!current.next) {
        const item = makeItem(value);
        item.next = current.next;
        current.next = item;
        return head;
      }

      current = current.next;
    }

    return head;
  }

  function dequeue() {
    const value = head.value;
    head = head.next;
    return value;
  }

  function printList() {
    let item = head;
    while (item) {
      console.log("%d", item.value);
      item = item.next;
    }
  }

  return {
    enqueue,
    dequeue,
    printList
  }
}

export default PQ;

// const pq = PQ((a, b) => a.value < b.value);

// pq.enqueue({value: 45});
// pq.enqueue({value: 15});
// pq.enqueue({value: 5});
// pq.enqueue({value: 89});
// pq.enqueue({value: 14});
// pq.enqueue({value: 16});
// // // pq.printList();

// console.log(pq.dequeue())
// console.log(pq.dequeue())
// console.log(pq.dequeue())




